File: app/src/main/java/de/rwth_aachen/phyphox/Bluetooth/BluetoothOutput.java
package de.rwth_aachen.phyphox.Bluetooth;

import android.app.Activity;
import android.bluetooth.BluetoothGattCharacteristic;
import android.content.Context;
import android.os.Build;

import androidx.annotation.RequiresApi;

import java.util.UUID;
import java.util.Vector;

import de.rwth_aachen.phyphox.DataBuffer;
import de.rwth_aachen.phyphox.DataInput;


/**
 * The BluetoothOutput class encapsulates an output to Bluetooth devices.
 */
@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
public class BluetoothOutput extends Bluetooth {

    /**
     * Data-buffers
     */
    public Vector<DataInput> data;

    /**
     * Create a new BluetoothOutput.
     *
     * @param idString        An identifier given by the experiment author used to group multiple devices and allow the user to distinguish them
     * @param deviceName      name of the device (can be null if deviceAddress is not null)
     * @param deviceAddress   address of the device (can be null if deviceName is not null)
     * @param uuidFilter       Optional filter to identify devices by advertices service or attribute UUIDs
     * @param autoConnect     If true, phyphox will not show a scan dialog and connect with the first matching device instead
     * @param context         context
     * @param buffers         list of dataOutputs to write the values
     * @param characteristics list of all characteristics the object should be able to operate on
     */
    public BluetoothOutput(String idString, String deviceName, String deviceAddress, UUID uuidFilter, Boolean autoConnect, Activity activity, Context context, Vector<DataInput> buffers, Vector<CharacteristicData> characteristics) {

        super(idString, deviceName, deviceAddress, uuidFilter, autoConnect, activity, context, characteristics);

        this.data = buffers;
    }


    /**
     * Write data to the Characteristics.
     */
    public void sendData() {
        if (!forcedBreak) {
            for (BluetoothGattCharacteristic characteristic : mapping.keySet()) {
                for (Characteristic c : mapping.get(characteristic)) {
                    if (data.get(c.index).getFilledSize() != 0) {
                        byte[] value = convertData(data.get(c.index).buffer, c.outputConversionFunction);
                        characteristic.setValue(value);
                        add(new WriteCommand(btGatt, characteristic));
                    }
                }
            }
        }
    }

    /**
     * Convert data using the specified conversion function.
     * Return a byte array with size 1 in case of an exception.
     *
     * @param data               data that should be converted
     * @param conversionFunction method to convert data (from ConversionsOutput)
     * @return the converted value
     */
    private byte[] convertData(DataBuffer data, ConversionsOutput.OutputConversion conversionFunction) {
        try {
            return conversionFunction.convert(data);
        } catch (Exception e) {
            return new byte[0]; // the method needs to return a byte array
        }
    }

} // end of class BluetoothOutput

File: app/src/main/java/de/rwth_aachen/phyphox/Bluetooth/BluetoothOutput.kt
package de.rwth_aachen.phyphox.Bluetooth

import android.app.Activity
import android.content.Context
import de.rwth_aachen.phyphox.DataBuffer
import de.rwth_aachen.phyphox.DataInput
import java.util.UUID
import java.util.Vector

/** A BLE device that receives buffer data after each analysis cycle or on a button trigger. */
class BluetoothOutput(
    idString: String?,
    deviceName: String?,
    deviceAddress: String?,
    uuidFilter: UUID?,
    autoConnect: Boolean,
    activity: Activity,
    context: Context,
    @JvmField
    val data: Vector<DataInput>,
    characteristics: Vector<CharacteristicData>
) : Bluetooth(idString, deviceName, deviceAddress, uuidFilter, autoConnect, activity, context, characteristics) {

    private val requestedTriggers = HashSet<String>()

    fun requestSend(triggerId: String) {
        requestedTriggers.add(triggerId)
    }

    // Queued writes coalesce: an untransmitted value is replaced by the newer one
    fun sendData() {
        if (forcedBreak)
            return

        for ((characteristic, characteristicList) in mapping) {
            var n = 0
            var out: ByteArray? = null
            for (c in characteristicList) {
                val triggerId = c.triggerId
                if (!triggerId.isNullOrEmpty() && !requestedTriggers.contains(triggerId))
                    continue

                if (data[c.index].filledSize != 0) {
                    val value = convertData(data[c.index].buffer, c.outputConversionFunction)
                    val offset = c.outputOffset.toInt()
                    if (value.size + offset > n) {
                        n = value.size + offset
                        val newOut = ByteArray(n)
                        out?.copyInto(newOut)
                        out = newOut
                    }
                    value.copyInto(out!!, offset)
                    if (!data[c.index].keep)
                        data[c.index].clear(false)
                }
            }
            if (out != null) {
                submitDataWrite(characteristic.uuid, out)
            }
        }
        requestedTriggers.clear()
    }

    private fun convertData(data: DataBuffer?, conversionFunction: ConversionsOutput.OutputConversion?): ByteArray {
        return try {
            conversionFunction?.convert(data) ?: ByteArray(0)
        } catch (e: Exception) {
            ByteArray(0)
        }
    }
}
